distbench_engine.cc

```diff
@@ -660,14 +685,29 @@ void DistBenchEngine::RunActionList(int list_index,
           ((size == 1) ||
            s.state_table[i].action->proto.send_response_when_done())) {
         sent_response_early = true;
-        s.state_table[i].all_done_callback = [&s, i, incoming_rpc_state]() {
+        s.state_table[i].all_done_callback = [&s, i, incoming_rpc_state,
+                                              this]() {
           incoming_rpc_state->SendResponseIfSet();
+          canceled_mu_.Lock();
```

Comment:
This could be
if (s.state_table[i].action->proto.cancel_traffic_when_done()) {
  TryToCancelTraffic()
}

And inside that helper function you can avoid grabbing the mutex by checking once and then again after grabbing it.